<a href="https://colab.research.google.com/github/100485446-UC3M/Primera_Practica_Machine_Learning/blob/main/Predicciones.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predicciones Práctica 1
## Esteban Gómez Buitrago 100485446
## Nicolás Alejandro Cuesta García 100495966

## Análisis Exploratorio de Datos

PAra realizar este projecto utilizamos los set de datos "attrition_availabledata_10.csv" y"attrition_competition_10.csv". Decidimos utilizar estos, después de realizar la suma de los últimos dos digitos del NIA 100485446 (4 + 6 = 10).

Al analizar los datos, la primera conclusión a la que se puede llegar es que este es un problema de clasificación. Se está realizando un estudio en el cual a través de algunas variables se va a definir si la clase (en este caso, el attributo "Attrition") es "Yes" o "No".

Para encontrar ciertos datos importantes para continuar con el análisis, será necesario utilizar ciertas funciones de python. El código y los resultados están descritos a continuación:

In [ ]:
#Las librerías necesarias son importadas
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [34]:
#Se crea una semilla para poder replicar los resultado y se leen los datos
np.random.seed(100495966)
data = pd.read_csv('attrition_availabledata_10.csv')

#Para chequear que sea el set de datos correcto
#print(data.head())

print(f"Número de variables (columnas): {data.shape[1] - 1}")
print(f"Número de instancias (filas): {data.shape[0]}")
print("\nLa cantidad de datos vacíos por columna es: ")
valoresvacios = data.isnull().sum()
print("Valores vacíos por columna:\n", valoresvacios.to_dict())
print(f"Número total de valores vacíos: {data.isnull().sum().sum()}")

#Las columnas de tipo object son las categóricas
categorical_cols = data.select_dtypes(include=['object'])
unique_counts = categorical_cols.nunique()
print("\nNúmero de categorías por variable categórica:\n", unique_counts.to_dict())
constant_columns = []
for column, count in unique_counts.items():
  if count == 1:
    constant_columns.append(column)
print("\nColumnas constantes:\n", constant_columns)
id_cols = data.columns[data.nunique() == data.shape[0]]
print("Columnas de ID:\n", id_cols)

num_yes = data["Attrition"].eq("Yes").sum()
num_no = data["Attrition"].eq("No").sum()
Total_attrition = num_no + num_yes
if Total_attrition == data.shape[0]:
  print(f"\nNúmero de empleados que dejaron la empresa (Attrition = Yes): {num_yes}")
  print(f"Número de empleados que no dejaron la empresa (Attrition = No): {num_no}")
else:
  print("Error en el número de instancias, hay valores no reconocidos")


Número de variables (columnas): 30
Número de instancias (filas): 2940

La cantidad de datos vacíos por columna es: 
Valores vacíos por columna:
 {'hrs': 0, 'absences': 0, 'JobInvolvement': 0, 'PerformanceRating': 0, 'EnvironmentSatisfaction': 15, 'JobSatisfaction': 12, 'WorkLifeBalance': 29, 'Age': 0, 'BusinessTravel': 0, 'Department': 0, 'DistanceFromHome': 0, 'Education': 0, 'EducationField': 0, 'EmployeeCount': 0, 'EmployeeID': 0, 'Gender': 0, 'JobLevel': 0, 'JobRole': 0, 'MaritalStatus': 0, 'MonthlyIncome': 0, 'NumCompaniesWorked': 17, 'Over18': 0, 'PercentSalaryHike': 0, 'StandardHours': 0, 'StockOptionLevel': 0, 'TotalWorkingYears': 5, 'TrainingTimesLastYear': 0, 'YearsAtCompany': 0, 'YearsSinceLastPromotion': 0, 'YearsWithCurrManager': 0, 'Attrition': 0}
Número total de valores vacíos: 78

Número de categorías por variable categórica:
 {'BusinessTravel': 3, 'Department': 3, 'EducationField': 6, 'Gender': 2, 'JobRole': 9, 'MaritalStatus': 3, 'Over18': 1, 'Attrition': 2}

Columnas

El estudio será realizado con 30 variables y 2940 instancias distintas. Es importante resaltar que al número de variables se le resto la variable "Attrition" ya que esta es nuestra variable de calisifcación (objetivo). Después de analizar las variable individualmente encontramos que tenemos variables catégoricas, ordinales y númericas.

Las variabes **catégoricas** son (7):   

*   BusinessTravel
*   Department
*   EducationField
*   Gender
*   JobRole
*   MaritalStatus
*   Over18

Las **númericas** incluyen (15):  

*   hrs
*   absences
*   Age
*   DistanceFromHome
*   MonthlyIncome
*   PercentSalaryHike
*   TotalWorkingYears
*   TrainingTimesLastYear
*   NumCompaniesWorked
*   EmployeeCount
*   EmployeeID
*   StandardHours
*   YearsAtCompany
*   YearsSinceLastPromotion
*   YearsWithCurrManager

Finalmente, entre las **ordinales** están (8):

*   JobInvolvement
*   JobLevel
*   Education
*   PerformanceRating
*   EnvironmentalSatisfaction
*   JobSatisfaction
*   WorkLifeBalance
*   StockOptionLevel

Para definir si una variable es de **alta cardinalidad** o no vamos a tomar el límite de **5 categorías**. Tomando este límite podemos ver que **existen 2 variables que tendrían alta cardinalidad**: EducationField (6) y JobRole (9). El resto de variables categóricas tienen 3 o menos categorías cada una, por lo que no tienen una cardinalidad alta.

Asimismo, es pertinente analizar si tenemos variables que tengan el mismo valor en todas las instancia (**columnas constantes**).  Como se puede evidenciar en el código anterior, exite una variable "Over18", que siempre tiene el mismo valor ("Y"). De esta forma se puede concluir que **"Over18" es una columna constante** y no le va a proporciona información útil al modelo, se deberá tratar con ella en el preproceso.

Hay que buscar **columnas de ID** que son columnas que tienen un valor diferente para cada instancia, y pueden servir como identificadores. Con el código encontramos que tenemos una columna que cumple con esta descripción. La variable **"EmployeeID" es una columna de ID**.

Entre todas estas instancias tenemos **78 valores vacíos** dividos en 5 variables: WorkLifeBalance (29),  NumCompaniesWorked: (17), EnvironmentSatisfaction (15), Job Satisfaction (12), y TotalWorkingYears (5). Esto es un factor a tener en cuenta a la hora de realizar el preproceso.

Cómo ya se describió antes, este es un problema de **clasificación**, para  finalizar, es pertinente analizar si está balanceado. La variable "*Attrition*" es nuestro clasificador, esta puede tener el valor de "Yes" o  "No". Para saber si está balanceado se debe ver cuantas intancias existen con cada valor. Como se puede ver en el código expuesto previamente, en este caso exiten 474 instancias que tienen en esta variable el valor de  "Yes" y 2466 que tienen "No". Con esto se puede concluir que el problema **NO está balanceado**, y se tendrá que utilizar una técnica de balanceo durante el preproceso.
